src/js/stores/document.js

```diff
@@ -166,18 +173,37 @@ define(function (require, exports, module) {
             if (suppressChange) {
                 return;
             }
-
-            // If some selected layer remains uninitialized, a new document will
-            // come along shortly. Wait until then to trigger the change event.
-            var initialized = nextDocument.layers && nextDocument.layers.selected
-                .every(function (layer) {
-                    return layer.initialized;
-                });
-
-            if (initialized) {
-                // Include the document in the event payload
-                this.emit(changeEventName || "change", nextDocument);
+            
+            var promise = Promise.bind(this);
+
+            // Notify change of layer selection state.
+            if (oldDocument) {
+                promise = promise.then(function () {
+                    oldDocument.layers.layers.forEach(function (layer) {
```

Comment:
Is this iteration slow on large documents? It always worries me when we iterate over all layers. Would it be worth it to try to optimize this by uninstalling listeners on layers that are, e.g., hidden in the layers panel? (That could well be over-optimization, but hey who knows.)